### PR TITLE
[#2110] Add Device DAO for accessing device collection

### PR DESCRIPTION
**Pull request by sophokles73**
This is the second step of refactoring the MongoDB based registry implementation (#2110).
A final PR will introduce a DAO for accessing credentials data.

### CODE REVIEW

**Comment by sophokles73** [comment]
@kaniyan would you mind taking a look if you find the time?

**Comment by kaniyan** [comment]
> @kaniyan would you mind taking a look if you find the time?

Sure I will have a look.

File: services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDao.java [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.UpdateOptions;
+
+
+/**
+ * A data access object for persisting device information to a Mongo DB collection.
+ *
+ */
+public class MongoDbBasedDeviceDao extends MongoDbBasedDao implements DeviceDao, HealthCheckProvider {
```
Comment (kaniyan) [inline services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDao.java:63]: How about making this class _final_?

**Review by kaniyan — CHANGES_REQUESTED** [review]

File: services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDao.java [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+    private static List<DeviceWithId> getDevicesWithId(final JsonObject searchResult) {
+        return Optional.ofNullable(searchResult.getJsonArray(RegistryManagementConstants.FIELD_RESULT_SET_PAGE))
+                .map(devices -> devices.stream()
+                        .filter(JsonObject.class::isInstance)
+                        .map(JsonObject.class::cast)
+                        .map(json -> {
+                            LOG.debug("document: {}{}", System.lineSeparator(), json.encodePrettily());
+                            return json;
+                        })
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedDeviceDao.java:345]: nice :+1:

**Comment by sophokles73** [comment]
@kaniyan thanks for the review :+1:  I have pushed some changes ...

**Review by kaniyan — APPROVED** [review]
LGTM

**Pull request merged**